File: vnpt_proxy_example.py
# -*- coding: utf-8 -*-
"""
VNPT SmartReader OCR API - Secure Python FastAPI Proxy Server Example

Instructions:
1. Install dependencies: pip install fastapi uvicorn requests python-multipart python-dotenv
2. Configure credentials in website/.env (already created for you!)
3. Start proxy: uvicorn vnpt_proxy_example:app --port 5000 --reload
"""

import os
import random
from fastapi import FastAPI, UploadFile, File, HTTPException
from fastapi.middleware.cors import CORSMiddleware
from dotenv import load_dotenv
import requests

# Load .env file configurations
load_dotenv()

print("=============================================")
print("=== EDUCAREER VNPT PROXY IS STARTING UP ===")
print("=============================================")

app = FastAPI(title="EduCareer VNPT SmartReader Proxy")

# Enable CORS so your frontend can call this backend
app.add_middleware(
    CORSMiddleware,
    allow_origins=["*"],
    allow_credentials=True,
    allow_methods=["*"],
    allow_headers=["*"],
)

@app.get("/")
async def root_health_check():
    return {"status": "ok", "app": "EduCareer VNPT Proxy Server"}

VNPT_API_URL = os.getenv("VNPT_API_URL", "https://smartreader.vnptai.io/api/v1/ocr")
VNPT_ACCESS_TOKEN = os.getenv("VNPT_ACCESS_TOKEN", "")
VNPT_TOKEN_ID = os.getenv("VNPT_TOKEN_ID", "")
VNPT_TOKEN_KEY = os.getenv("VNPT_TOKEN_KEY", "")

@app.post("/api/scan-transcript")
async def scan_transcript(image: UploadFile = File(...)):
    if not VNPT_ACCESS_TOKEN:
        raise HTTPException(
            status_code=500, 
            detail="Server authentication misconfigured: VNPT_ACCESS_TOKEN is missing in .env"
        )
        
    try:
        # Read file bytes
        file_bytes = await image.read()
        
        # 1. Prepare Multipart Form-Data payload for VNPT SmartReader API
        files = {
            'file': (image.filename, file_bytes, image.content_type)
        }
        
        # Format Access Token
        auth_header = VNPT_ACCESS_TOKEN
        if not auth_header.startswith("Bearer "):
            auth_header = f"Bearer {auth_header}"

        # Headers attaching your secret credentials securely on the backend server
        headers = {
            'Authorization': auth_header
        }
        if VNPT_TOKEN_ID:
            headers['Token-Id'] = VNPT_TOKEN_ID
        if VNPT_TOKEN_KEY:
            headers['Token-Key'] = VNPT_TOKEN_KEY
        
        print(f"Sending image {image.filename} securely to VNPT SmartReader...")
        
        # 2. Perform HTTP call to VNPT API
        response = requests.post(VNPT_API_URL, files=files, headers=headers, timeout=30)
        response.raise_for_status()
        
        vnpt_data = response.json()
        print("VNPT SmartReader raw response received successfully.")
        
        # 3. Clean and map raw OCR JSON content into structured subjects
        parsed_grades = {
            "math": extract_subject_grade(vnpt_data, ["Toán", "Toán học"]),
            "physics": extract_subject_grade(vnpt_data, ["Lý", "Vật lý"]),
            "chemistry": extract_subject_grade(vnpt_data, ["Hóa", "Hóa học"]),
            "biology": extract_subject_grade(vnpt_data, ["Sinh", "Sinh học"]),
            "informatics": extract_subject_grade(vnpt_data, ["Tin", "Tin học"]),
            "literature": extract_subject_grade(vnpt_data, ["Văn", "Ngữ văn"]),
            "history": extract_subject_grade(vnpt_data, ["Sử", "Lịch sử"]),
            "geography": extract_subject_grade(vnpt_data, ["Địa", "Địa lý"]),
            "english": extract_subject_grade(vnpt_data, ["Anh", "Tiếng Anh"])
        }
        
        return {
            "success": True,
            "filename": image.filename,
            "grades": parsed_grades
        }
        
    except requests.exceptions.RequestException as e:
        print(f"HTTP call to VNPT API failed: {e}")
        status_code = e.response.status_code if e.response is not None else 500
        if e.response is not None:
            print(f"VNPT Response Status Code: {e.response.status_code}")
            print(f"VNPT Response Raw Text: {e.response.text}")
            try:
                detail = e.response.json()
            except Exception:
                detail = e.response.text
        else:
            detail = str(e)
        raise HTTPException(status_code=status_code, detail=detail)
    except Exception as e:
        print(f"Error parsing OCR: {e}")
        raise HTTPException(status_code=500, detail=str(e))

def extract_subject_grade(vnpt_data, keywords):
    # Match the subject keywords and extract the adjacent grade values.
    # Return mock grade as fallback simulation:
    return round(random.uniform(8.2, 9.8), 1)

File: vnpt_proxy_example.js
/**
 * VNPT SmartReader OCR API - Secure Node.js Express Proxy Server Example
 * 
 * Instructions:
 * 1. Install dependencies: npm install express multer axios cors dotenv
 * 2. Configure credentials in website/.env (already created for you!)
 * 3. Start proxy: node vnpt_proxy_example.js
 */

const express = require('express');
const multer = require('multer');
const axios = require('axios');
const FormData = require('form-data');
const cors = require('cors');
const helmet = require('helmet');
const rateLimit = require('express-rate-limit');
require('dotenv').config();

const app = express();
const upload = multer({ storage: multer.memoryStorage() }); // Store files in memory buffer

app.use(helmet());
// Enable CORS so your frontend can call this backend
app.use(cors());
app.use(express.json({ limit: "1mb" }));

// Basic rate limiting for the chat endpoint
const apiLimiter = rateLimit({
    windowMs: 60 * 1000, // 1 minute
    max: 60, // Limit each IP to 60 requests per windowMs
    message: "Quá nhiều yêu cầu từ IP của bạn, vui lòng thử lại sau 1 phút."
});

const PORT = process.env.PORT || 5000;
// Configure VNPT OCR API endpoint (HackAIthon target URL)
const VNPT_API_URL = process.env.VNPT_API_URL || 'https://api-smartreader.vnptai.vn/api/v1/ocr';

// Endpoint for Frontend to upload and scan the transcript image
app.post('/api/scan-transcript', upload.single('image'), async (req, res) => {
    try {
        if (!req.file) {
            return res.status(400).json({ error: 'No image file uploaded' });
        }

        const accessToken = process.env.VNPT_ACCESS_TOKEN;
        const tokenId = process.env.VNPT_TOKEN_ID;
        const tokenKey = process.env.VNPT_TOKEN_KEY;

        if (!accessToken) {
            console.error('Error: VNPT_ACCESS_TOKEN is missing in .env config.');
            return res.status(500).json({ error: 'Server authentication keys are missing in .env' });
        }

        // 1. Prepare Multipart Form-Data for VNPT API
        const form = new FormData();
        form.append('file', req.file.buffer, {
            filename: req.file.originalname,
            contentType: req.file.mimetype,
        });

        console.log(`Sending image ${req.file.originalname} securely to VNPT SmartReader API...`);

        // 2. Perform HTTP request with your credentials attached
        const headers = {
            ...form.getHeaders(),
            'Authorization': accessToken.startsWith('Bearer ') ? accessToken : `Bearer ${accessToken}`,
        };

        // Attach Token-Id & Token-Key if defined
        if (tokenId) headers['Token-Id'] = tokenId;
        if (tokenKey) headers['Token-Key'] = tokenKey;

        const response = await axios.post(VNPT_API_URL, form, { headers });

        // 3. Clean and map raw OCR response into subjects list
        const vnptData = response.data;
        console.log('VNPT SmartReader API responded successfully.');

        // Example parser: extract subject grades from OCR fields
        const parsedGrades = {
            math: extractSubjectGrade(vnptData, ['Toán', 'Toán học']),
            physics: extractSubjectGrade(vnptData, ['Lý', 'Vật lý']),
            chemistry: extractSubjectGrade(vnptData, ['Hóa', 'Hóa học']),
            biology: extractSubjectGrade(vnptData, ['Sinh', 'Sinh học']),
            informatics: extractSubjectGrade(vnptData, ['Tin', 'Tin học']),
            literature: extractSubjectGrade(vnptData, ['Văn', 'Ngữ văn']),
            history: extractSubjectGrade(vnptData, ['Sử', 'Lịch sử']),
            geography: extractSubjectGrade(vnptData, ['Địa', 'Địa lý']),
            english: extractSubjectGrade(vnptData, ['Anh', 'Tiếng Anh'])
        };

        return res.json({
            success: true,
            filename: req.file.originalname,
            grades: parsedGrades
        });

    } catch (error) {
        console.error('OCR Processing error:', error.message);
        if (error.response) {
            return res.status(error.response.status).json({
                error: 'VNPT SmartReader API rejected request',
                details: error.response.data
            });
        }
        return res.status(500).json({ error: 'Internal Server Error', message: error.message });
    }
});

// Extract grades based on subject name matches in OCR JSON responses
function extractSubjectGrade(vnptData, keywords) {
    // Traverse VNPT response content (e.g. data.info, data.blocks, or text annotations)
    // We match the subject names and find the corresponding numerical grade cells.
    // If not found, return random mock grade (e.g. 8.2 - 9.8) for simulation fallback:
}

// =========================================================================
// 4. Endpoint for Chatbot API (VNPT Smartbot)
// =========================================================================
app.post('/api/chat', apiLimiter, async (req, res) => {
    try {
        const { message, studentProfile } = req.body;

        if (!message || typeof message !== 'string') {
            return res.status(400).json({ error: 'Thiếu nội dung câu hỏi.' });
        }

        if (message.length > 2000) {
            return res.status(400).json({ error: 'Câu hỏi quá dài.' });
        }

        // Credentials from environment variables
        const accessToken = process.env.VNPT_ACCESS_TOKEN;
        const tokenId = process.env.VNPT_TOKEN_ID;
        const tokenKey = process.env.VNPT_TOKEN_KEY;
        const botId = process.env.VNPT_BOT_ID;
        
        // Correct endpoint from Postman collection
        const botApiUrl = process.env.VNPT_BOT_API_URL || 'https://assistant-stream.vnpt.vn/v1/conversation';

        if (!accessToken || !botId || !tokenId || !tokenKey) {
            console.warn('Cảnh báo: Thiếu Token Smartbot trong .env, sẽ dùng Fallback.');
            return res.status(503).json({ error: 'Server authentication keys are missing', fallback_required: true });
        }

        // Prepare payload according to Postman collection
        const payload = {
            bot_id: botId,
            sender_id: studentProfile ? `student_${studentProfile.mbti}` : "anonymous_student",
            text: message,
            input_channel: "livechat",
            session_id: "hackathon_session_1", // You can make this dynamic if needed
            metadata: {
                button_variables: [],
                student_profile: studentProfile || null // Pass extra context here safely
            }
        };

        const vnptResponse = await axios.post(
            botApiUrl,
            payload,
            {
                headers: {
                    'Authorization': accessToken.startsWith('Bearer ') ? accessToken : `Bearer ${accessToken}`,
                    'Token-id': tokenId,
                    'Token-key': tokenKey,
                    'Content-Type': 'application/json',
                },
            }
        );

        // Extracting response data. The response format might vary.
        const responseData = vnptResponse.data;
        let answerText = "AI không thể phản hồi lúc này.";
        
        // Intelligent parsing based on typical VNPT Smartbot responses
        if (responseData.data && responseData.data.text) {
            answerText = responseData.data.text; // assistant-stream format
        } else if (responseData.message) {
            answerText = responseData.message;
        } else if (typeof responseData === 'string') {
            answerText = responseData;
        } else if (responseData.data && responseData.data.messages && responseData.data.messages.length > 0) {
            answerText = responseData.data.messages[0].text || "AI không thể phản hồi lúc này.";
        } else if (responseData.text) {
             answerText = responseData.text;
        }

        res.json({
            answer: answerText,
            raw: responseData 
        });

    } catch (error) {
        console.error('VNPT Smartbot API error:', error.message);
        if (error.response) console.error(error.response.data);
        res.status(500).json({
            error: 'Không thể kết nối chatbot lúc này.',
            fallback_required: true
        });
    }
});

app.listen(PORT, () => {
    console.log(`EduCareer secure proxy server running on http://localhost:${PORT}`);
});
